flaky e2e: iceberg_source_position_delete.slt position-delete assertion races the checkpoint boundary

## Describe the bug

The e2e test `e2e_test/iceberg/test_case/iceberg_source_position_delete.slt` is flaky: its PositionDeletes assertion races the sink checkpoint boundary and can fail under CI load, and when it fails it additionally poisons the shared `demo_db` namespace, cascading into failures of unrelated test cases in the same shard.

Observed failure (Buildkite `pull-request` build 98070, job "end-to-end iceberg test", shard 2/2):
https://buildkite.com/risingwavelabs/pull-request/builds/98070#019f3dad-226e-4158-b7fd-1677063bac62

```
test_case/iceberg_source_position_delete.slt .. [FAILED] after 21975 ms

failed to run `test_case/iceberg_source_position_delete.slt`

Caused by:
    query result mismatch:
    [SQL] select count(*) >= 1 from rw_iceberg_files where content = 'PositionDeletes';
    [Diff] (-expected|+actual)
    -   t
    +   f
    at test_case/iceberg_source_position_delete.slt:57
```

### Cascading failures

Because the test aborted at line 57, its trailing cleanup (`DROP SINK` / `DROP SOURCE` / `DROP TABLE ... cascade`) never ran, leaving the `test_position_delete` table behind in `demo_db`. The next three test cases in the same shard then failed on setup:

```
[ERROR] test case .../no_partition_append_only.toml failed: (org.apache.iceberg.exceptions.NamespaceNotEmptyException) Namespace demo_db is not empty.
[ERROR] test case .../partition_append_only.toml failed: (org.apache.iceberg.exceptions.NamespaceNotEmptyException) Namespace demo_db is not empty.
[ERROR] test case .../range_partition_append_only.toml failed: (org.apache.iceberg.exceptions.NamespaceNotEmptyException) Namespace demo_db is not empty.
```

So one flaky assertion produced 4 red test cases (`Total Tests: 10, Passed: 6, Failed: 4`).

## Root cause

The test (added in #23591) relies on a timing property that is not guaranteed:

1. It creates an upsert iceberg sink with sink decoupling (commit roughly every 10s, `commit_checkpoint_interval = 5`), inserts 5 rows, `flush`es, then `DELETE FROM s1 WHERE i1 < 3`, sleeps 20s, and asserts `rw_iceberg_files` contains at least one `PositionDeletes` file.
2. Whether a delete becomes a **position delete** or an **equality delete** is decided by the iceberg upsert delta writer based on the write session: a delete of a row written **in the same commit window** becomes a position delete against the just-written data file; a delete of a row **already committed** in a prior checkpoint becomes an equality delete.
3. Under CI load, a checkpoint/commit boundary can land between the `INSERT ... flush` and the `DELETE`. The 5 rows are then already committed when the delete arrives, so the sink emits only an equality-delete file, no position-delete file — and the assertion at line 57 sees `f`.

`rw_iceberg_files.content` (`src/frontend/src/catalog/system_catalog/rw_catalog/rw_iceberg_files.rs`) reads directly from the live manifest entries of the current snapshot, so the `f` result faithfully reflects that no position-delete file was written — the flakiness is in the test's timing assumption, not in the catalog or the source read path.

Note: the failure fired on PR #26080 (VARIANT type), but it is unrelated to that PR — the branch touches nothing under `src/connector/src/sink/iceberg/` (the position/equality-delete write path), and the test uses only `int`/`varchar` columns. The job log contains no VARIANT-related errors.

## Expected behavior

The test should pass deterministically regardless of where the checkpoint boundary falls, and a failure in one test case should not cascade into `NamespaceNotEmptyException` failures of subsequent test cases.

## Suggested fixes

1. **Make the assertion deterministic**: ensure the INSERT and DELETE land in the same commit window — e.g. drop the intervening `flush` so both changes ride the same checkpoint, or otherwise gate the sink commit until after the DELETE is issued. Alternatively, relax/rewrite the assertion so it doesn't depend on the delete-file flavor.
2. **Make cleanup robust**: the test harness (or per-case `drop_sqls`/init) should drop leftover tables in `demo_db` even when the previous `.slt` aborted mid-file, so one failure doesn't poison the namespace for the rest of the shard.

## How was this found

CI triage of Buildkite `pull-request` build 98070 on commit `d5d0e8b3523aee37c2a5c62b433f768350915e07` (branch `dylan/support_variant_type`). The sibling shard (1/2) of the same build passed; it runs a disjoint test set that does not include this test.

--- a/e2e_test/iceberg/main.py
+++ b/e2e_test/iceberg/main.py
@@ -28,6 +28,7 @@ def run_test_case(test_file: str, args: dict) -> TestResult:
     start_time = datetime.now()
     error = None
     success = True
+    drop_sqls = None
 
     try:
         with open(test_file, "rb") as f:
@@ -55,8 +56,6 @@ def run_test_case(test_file: str, args: dict) -> TestResult:
                 compare_sql(args, cmp_sqls)
             if verify_slt:
                 execute_slt(args, verify_slt)
-            if drop_sqls:
-                drop_table(args, drop_sqls)
     except KeyboardInterrupt:
         log(f"test case {test_file} interrupted by user", level=LogLevel.ERROR)
         error = "Interrupted by user"
@@ -70,6 +69,16 @@ def run_test_case(test_file: str, args: dict) -> TestResult:
         success = False
         raise e
     finally:
+        # Clean up even when the case failed: a leftover table keeps the shared
+        # namespace non-empty and fails every later case's DROP SCHEMA.
+        if drop_sqls:
+            try:
+                drop_table(args, drop_sqls)
+            except Exception as e:
+                # drop_table already logged the cause; don't mask an earlier failure.
+                if success:
+                    error = e
+                    success = False
         duration = datetime.now() - start_time
         return TestResult(test_file, success, duration, error)
 

--- a/e2e_test/iceberg/test_case/iceberg_source_position_delete.slt
+++ b/e2e_test/iceberg/test_case/iceberg_source_position_delete.slt
@@ -1,4 +1,7 @@
-# Deletions in a single commit are posistion delete, deletions across multiple commits are equail delete. sink_decouple = default(true), so we'll commit every 10s.
+# A delete is encoded as a position delete only when it shares an iceberg commit with the
+# insert it deletes, and as an equality delete otherwise. The sink commits every
+# commit_checkpoint_interval checkpoints counted from sink startup, so the interval has to be
+# comfortably larger than the number of checkpoints the statements below can consume.
 statement ok
 set streaming_parallelism=4;
 
@@ -22,7 +25,7 @@ CREATE SINK sink1 AS select * from mv1 WITH (
     s3.access.key = 'hummockadmin',
     s3.secret.key = 'hummockadmin',
     create_table_if_not_exists = 'true',
-    commit_checkpoint_interval = 5,
+    commit_checkpoint_interval = 20,
     primary_key = 'i1,i2',
     partition_by = 'i1'
 );
@@ -36,8 +39,10 @@ flush
 statement ok
 DELETE FROM s1 WHERE i1 < 3;
 
-sleep 20s
-
+# DELETE does not flush implicitly; flushing keeps it in the checkpoint right after the
+# insert's instead of leaving it to the next periodic barrier.
+statement ok
+flush
 
 statement ok
 CREATE SOURCE iceberg_t1_source
@@ -53,9 +58,9 @@ WITH (
     table.name = 'test_position_delete',
 );
 
-# check there should be at least one PositionDeletes
-query T
-select count(*) >= 1 from rw_iceberg_files where content = 'PositionDeletes';
+# Wait for the sink's first commit instead of sleeping for a fixed guess of the interval.
+query T retry 20 backoff 3s
+select count(*) >= 1 from rw_iceberg_files where source_name = 'iceberg_t1_source' and content = 'PositionDeletes';
 ----
 t
 

